plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,142 @@
+    | Pure a
+    deriving (Functor)
+
+instance Alt (Contract i o) where
+    (<!>) = select
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract i o) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> fmap f ca
+        Waiting f ->
+            case ca of
+                Emit t c -> Emit t (Waiting f <*> c)
+                Pure a' -> Waiting $ \i' -> fmap (\f' -> f' a') (f i')
+                Waiting f' ->
+                    Waiting $ \i' -> f i' <*> f' i'
+
+-- The monad instance sequentialises the 'Waiting' operations
+instance Monad (Contract i o) where
+    c >>= f = case c of
+        Waiting f' -> Waiting (f' >=> f)
+        Pure a     -> f a
+        Emit t c'  -> Emit t (c' >>= f)
+
+instance Semigroup a => Semigroup (Contract i o a) where
+    (<>) = liftA2 (<>)
+
+-- | Run both contracts and return the first one that finishes
+select :: Contract i o a -> Contract i o a -> Contract i o a
+select l r = 
+    case l of
+        Pure a -> Pure a
+        Emit e c -> Emit e (select c r)
+        Waiting f -> 
+            go r where
+                go = \case
+                    Pure a -> Pure a
+                    Emit e c -> Emit e (go c)
+                    Waiting f' -> 
+                        Waiting $ \i -> select (f i) (f' i)
+
+
+emit :: o -> Contract i o ()
+emit t = Emit t (pure ())
+
+waiting :: Contract i o i
+waiting = Waiting pure
+
+-- https://hackage.haskell.org/package/extra-1.6.15/docs/src/Control.Monad.Extra.html#loopM
+
+-- | A monadic version of 'loop', where the predicate returns 'Left' as a seed for the next loop
+--   or 'Right' to abort the loop.
+loopM :: Monad m => (a -> m (Either a b)) -> a -> m b
+loopM act x = do
+    res <- act x
+    case res of
+        Left x' -> loopM act x'
+        Right v -> return v
+
+-- | Repeatedly evaluate the action until it yields 'Nothing',
+--   then return the aggregated result.
+foldMaybe 
+    :: Monad m 
+    => (a -> b -> b) 
+    -> b 
+    -> m (Maybe a) 
+    -> m b
+foldMaybe f b con = loopM go b where
+    go b' = maybe (Left b') (Right . flip f b') <$> con
+
+-- | Apply an input to a contract
+applyInput
+    :: i
+    -> Contract i o a
+    -> Contract i o a
+applyInput ip = \case
+    Waiting f -> f ip
+    Pure a -> Pure a -- ??
+    Emit t c -> Emit t (applyInput ip c)
+
+applyInputs
+    :: Contract i o a
+    -> [i]
+    -> Contract i o a
+applyInputs = foldl' (flip applyInput)
+
+drain :: Monoid o => Contract i o a -> (o, Contract i o a)
+drain = \case
+    Waiting f -> (mempty, Waiting f)
+    Pure a -> (mempty, Pure a)
+    Emit t c -> first (t <>) (drain c)
+
+outputs :: Monoid o => Contract i o a -> o
+outputs = fst . drain
+
+result :: Monoid o => Contract i o a -> Maybe a
+result = (\case { Pure b -> Just b; _ -> Nothing }) . snd . drain
+
+await :: o -> (i -> Maybe a) -> Contract i o a
```

Comment:
I remain suspicious about using this for triggers. As it is, all triggers must be set up in parallel, otherwise "downstream" triggers won't see any events, because they're discarded by earlier triggers looking at things. That seems like a very easy way to mess things up.

Reply:
> all triggers must be set up in parallel, otherwise "downstream" triggers won't see any events

That was the intention, the triggers aren't active all the time. If you use `select`, `until`, `both` etc. to build your contracts than it should work as expected.

Reply:
Hmm, so I think the situation I'm describing is something like (made up function names):
```
do
   _ <- waitUntilFundsAre x
   _ <- waitUntilSlotIs y
   doThing
````
Now, the question is: what happens if slot y passes and *then* funds reach x? I think if this was written in applicative style it would work, so probably using `ApplicativeDo` you're okay too. And if there is a data dependency (i.e. the second condition you're waiting on depends on the output of the first one, then of course it can't trigger until the first one does.

So I think this is indeed okay, although I'd like to see some tests of the basic mechanism witnessing the various patterns of awaiting etc.

Reply:
A general limitation of the current design is that there are no signals (in the FRP sense). So `waitUntilFundsAre` returns as soon as the funds are `x`, but then can't tell if the funds drop below `x` again by the time `waitUntilSlotIs` returns.

Although for your particular example you can use [`watchAddressUntil`](https://github.com/input-output-hk/plutus/pull/1013/files#diff-0c70358a7c647e9fb38c726c94108085R57).